File: Python/Code/svm.py
#! /usr/bin/env python

import sys, numpy, random
from sklearn.svm import SVC
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn import cross_validation

from lib import util, config

class SVM(object):

    def __init__(self):
        self.model = None

    def get_classifier(self):
        return SVC(C = 1, kernel = 'linear', class_weight = 'auto')

    def fit(self, x, y):
        self.model = self.get_classifier()
        self.model.fit(x, y)

    def score(self, x, y):
        if self.model is None:
            return None
        return numpy.mean(self.model.predict(x) == y)

def main():
    # initialize the data
    labels, stories = util.get_distress_data(config.CONNECTION)
    vec = TfidfVectorizer(ngram_range = (1, 5), strip_accents = None, charset_error = 'ignore', stop_words = None)
    instances = vec.fit_transform(stories)
    random.seed(0)

    # cross validate
    cv = 5; cv_accuracy = list();
    for i in xrange(0, cv):
        print "Iteration #" + str(i) + "..."

        # initialize training/testing data
        cv_data = cross_validation.train_test_split(instances, labels, test_size = 0.5, random_state = i)
        x_training = cv_data[0]
        x_testing = cv_data[1]
        y_training = cv_data[2]
        y_testing = cv_data[3]

        # initialize the classifier
        clf = SVM()
        clf.fit(x_training, y_training)

        # measure prediction accuracy
        cv_accuracy.append(clf.score(x_testing, y_testing))

    print "Scores => " + str(cv_accuracy)
    print "Mean   => " + str(numpy.mean(cv_accuracy))

if __name__ == "__main__":
    main()

File: MATLAB/Code/svm.m
clear; close all;

load('seeds.mat'); rng(s);

[labels, instances] = libsvmread('Data/n-gram.data');

cv = cvpartition(labels, 'HoldOut', 0.3);
cv_accuracy = zeros(1, cv.NumTestSets);
param = '-t 0 -c 1 -h 0 -w1 %.3f -w-1 %.3f';

for i = 1 : cv.NumTestSets
    fprintf('Iteration #%d\n', i);
    
    % initialize training/testing dataset
    training = cv.training(i);
    testing = cv.test(i);
    x_training = instances(training, :); y_training = labels(training, :);
    x_testing = instances(testing, :); y_testing = labels(testing, :);
    
    % train the model
    n = size(x_training, 1);
    w = ones(size(x_training, 1), 1);
    positive = size(y_training, 1) / sum(y_training == 1);
    negative = size(y_training, 1) / sum(y_training == -1);
    param = sprintf('-t 0 -c %s -w1 %.3f -w-1 %.3f', num2str(i * 10), positive, negative);
    model = svmtrain(w, y_training, x_training, param);

    % predict on the testing data
    [predictions, ~, ~] = svmpredict(y_testing, x_testing, model);
    
    cv_accuracy(i) = sum(y_testing == predictions) / size(y_testing, 1);
end

fprintf('Accuracy => [%s]\nMean => %s\n', num2str(cv_accuracy), num2str(mean(cv_accuracy)));
